Feat/483 fetcher options in project settings

luca-schlecker opened the pull request:
Closes #483 
Closes #489

## What I have made

- Update API to v0.6.0
- Add FetcherAddDialog
- Add FetcherOptionRow
- Add FetcherOptionsDialog
- Add FetcherRemovalDialog
- Add FetcherSettings
- Add UpdateFetchers
- Add E2E Tests for this feature

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does not apply.

- [x] I have updated the documentation accordingly and commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
  - [ ] unit tests
  - [x] integration tests
  - [x] end-to-end tests
- [x] (for reviewer) I have checked the implementation against the requirements

luca-schlecker committed:
chore: update api to v0.6.0

luca-schlecker committed:
feat: implement fetcher settings in project settings

domienderle reviewed (changes requested):
In general it looks good. I did not have a look on your code for now, because there are a couple of things, that have to be done first:
- If I add a fetcher, then click on `add fetcher` again and then instantly click on `add`, then I would excpect an error dialog, but I get a successfully added fetcher message and nothing happens. Please correct this, that the error dialog is correctly shown.
- If I once tried to add a fetcher without selecting one, then the error dialog is shown every time I click on `add fetcher`. I would excpect the error dialog to be gone until the next time I try to add a fetcher without selecting one.
- When I delete a fetcher, and click on `add fetcher` again, then the previously deleted fetcher is selected, whereas when I add a fetcher and click `add fetcher` again, no fetcher is selected. I would expect to don't pre-select a fetcher after I deleted one.
- change the button `add fetcher` to a loading button, that waits until all available fetchers are loaded (at the moment the dropdown for the fetcher is empty as long as the fetchers are not loaded)
- change the button `add` in the `FetcherAddDialog` to a loading button, that waits until the fetcher is added (at the moment you can click it multiple times until the dialog closes and get multiple toasts)
- those are only two examples. Please have a look on every button and make it a loading button if needed (the cancel buttons don't have to be loading buttons, but every other button, I guess, should be)
- What do you think of giving the user the option to add multiple fetchers at the same time? So, that I could select multiple fetchers in the dropdown. At the moment I have to click through the whole fetcher adding cycle for every fetcher I want to add. Especially for a newly created project, this could be very annoying. I would like to have this. What do you think?
- I would propose to make the names of the fields labels. Otherwise the user thinks, that he could edit the name

Additionally, I have a couple of questions:
- Does it make sense, that the `overriden` field is checked, if I add the default value (e.g. via click)? I would expect, that in this case nothing is overriden, because it still is the default value no matter if the text field is now filled or not, or do I get anything wrong?
- Does it have any specific ordering of the fields? Maybe you could sort the field names alphabetically. In this case every field appears to be on the same row for every fetcher. It is sort of weird to have for example the `API_KEY` sometimes as the first field and sometimes as the last field or in between.

luca-schlecker committed:
feat: add `loading` spinner to SettingsSection

luca-schlecker committed:
feat: add icon snippet to LoadingButton

luca-schlecker committed:
feat(FetcherSettings): add loading buttons & prefer label of readonly input

luca-schlecker commented:
> In general it looks good. I did not have a look on your code for now, because there are a couple of things, that have to be done first:
> 
> * If I add a fetcher, then click on `add fetcher` again and then instantly click on `add`, then I would excpect an error dialog, but I get a successfully added fetcher message and nothing happens. Please correct this, that the error dialog is correctly shown.
> * If I once tried to add a fetcher without selecting one, then the error dialog is shown every time I click on `add fetcher`. I would excpect the error dialog to be gone until the next time I try to add a fetcher without selecting one.
> * When I delete a fetcher, and click on `add fetcher` again, then the previously deleted fetcher is selected, whereas when I add a fetcher and click `add fetcher` again, no fetcher is selected. I would expect to don't pre-select a fetcher after I deleted one.
> * change the button `add fetcher` to a loading button, that waits until all available fetchers are loaded (at the moment the dropdown for the fetcher is empty as long as the fetchers are not loaded)
> * change the button `add` in the `FetcherAddDialog` to a loading button, that waits until the fetcher is added (at the moment you can click it multiple times until the dialog closes and get multiple toasts)
> * those are only two examples. Please have a look on every button and make it a loading button if needed (the cancel buttons don't have to be loading buttons, but every other button, I guess, should be)
> * What do you think of giving the user the option to add multiple fetchers at the same time? So, that I could select multiple fetchers in the dropdown. At the moment I have to click through the whole fetcher adding cycle for every fetcher I want to add. Especially for a newly created project, this could be very annoying. I would like to have this. What do you think?
> * I would propose to make the names of the fields labels. Otherwise the user thinks, that he could edit the name
> 
> Additionally, I have a couple of questions:
> 
> * Does it make sense, that the `overriden` field is checked, if I add the default value (e.g. via click)? I would expect, that in this case nothing is overriden, because it still is the default value no matter if the text field is now filled or not, or do I get anything wrong?
> * Does it have any specific ordering of the fields? Maybe you could sort the field names alphabetically. In this case every field appears to be on the same row for every fetcher. It is sort of weird to have for example the `API_KEY` sometimes as the first field and sometimes as the last field or in between.

These are great suggestions, thank you very much for the extensive test. I extended the `LoadingButton` so it can display an optional icon, and also added a loading spinner to `SettingsSection`. Furthermore, I added `LoadingButton`s to (hopefully all) loading actions in the fetcher settings. I also swapped out the read-only input with a label, just as you suggested.

Regarding your questions:
1. It is a finesse of how options are handled: The `overridden` column indicates if the option is actually sent to the fetcher. So even if you insert the displayed default, it is still in fact sent to the fetcher.
2. The order is dictated by the maps involved in the process. It is possible that you only noticed it that heavily because all fetchers in the mock backend share the same set of options, making a duplicate encounter statistically quite likely. I'd like to leave it like this for now and react in the future if it turns out to be an annoyance.

domienderle commented on src/lib/components/composites/button/LoadingButton.svelte:
Which use case does this have? I didn't see any icon in a loading spinner of this PR or am I wrong?

luca-schlecker replied:
The `Add Fetcher` button uses it.

domienderle replied:
Thanks for the clarification. I did not see this.

domienderle commented on src/lib/components/composites/settings/project-settings/slr/FetcherAddDialog.svelte:
```diff
@@ -0,0 +1,102 @@
+
+        loading = false;
+
+        fetchers = [];
+        error = undefined;
```
Do we need this, if we have the `effect` below?

luca-schlecker replied:
Nope, it caused one of the issues you've mentioned and has been removed.

domienderle commented on src/lib/components/composites/settings/project-settings/slr/FetcherAddDialog.svelte:
```diff
@@ -0,0 +1,102 @@
+    const content = $derived(
+        fetchers.length === 0
+            ? "Select a fetcher"
+            : `${fetchers.length} fetcher${fetchers.length > 0 ? "s" : ""} selected`,
```
This doesn't work. Also, we already have an extra method for this.
```suggestion
            : `${fetchers.length} ${pluralize(fetchers.length, "fetcher", "fetchers")} selected`,
```

luca-schlecker replied:
Oops. Off by one, should have been a `1`. Anyway, the pluralize function is now used.

domienderle commented on src/lib/components/composites/settings/project-settings/slr/FetcherOptionsDialog.svelte:
```diff
@@ -0,0 +1,156 @@
+<AlertDialog
+    actionButtonText="Save"
+    actionProps={{
+        onclick: () => saveFetcherOptions(),
```
```suggestion
        onclick: saveFetcherOptions,
```

domienderle commented on src/lib/components/composites/settings/project-settings/slr/FetcherOptionsDialog.svelte:
```diff
@@ -0,0 +1,156 @@
+    }}
+    cancelButtonText="Cancel"
+    cancelProps={{
+        onclick: () => {},
```
We don't need this line.
```suggestion

```

domienderle commented on src/lib/components/composites/settings/project-settings/slr/FetcherOptionsDialog.svelte:
```diff
@@ -0,0 +1,156 @@
+                        <Skeleton class="h-8 w-8" />
+                    </div>
+                {/each}
+            {:else if options.size == 0}
```
Use `===` instead of `==`.
```suggestion
            {:else if options.size === 0}
```

domienderle commented on src/lib/components/composites/settings/project-settings/slr/FetcherOptionsDialog.svelte:
```diff
@@ -0,0 +1,156 @@
+        );
+    }
+
+    let loading = $state(false);
```
Declare (`state`) variables at the top of the class.

domienderle commented on src/lib/components/composites/settings/project-settings/slr/FetcherOptionsDialog.svelte:
```diff
@@ -0,0 +1,156 @@
+
+    let loading = $state(false);
+
+    const headers: [string, string][] = [
```
Move `headers` also to the top of the class, to get a clearer structure.

domienderle commented on src/lib/components/composites/settings/project-settings/slr/FetcherOptionsDialog.svelte:
```diff
@@ -0,0 +1,156 @@
+            ),
+        };
+
+        updateFetchers(
```
Why don't we use `await` here?

luca-schlecker replied:
On accident. Resolved now.

domienderle commented on src/lib/components/composites/settings/project-settings/slr/FetcherOptionsDialog.svelte:
```diff
@@ -0,0 +1,156 @@
+    // Fetch the available options for a fetcher
+    async function refetchOptions(fetcher: string) {
+        loading = true;
+        const availableOptions = await backendService
+            .getAvailableFetcherOptions({ fetcherName: fetcher })
+            .then((it) => new Map(Object.entries(it.response.options)));
+        const fetchers = new Map(Object.entries(projectSettings?.fetchers ?? []));
+        const fetcherOptions = new Map(Object.entries(fetchers.get(fetcher)?.options ?? {}));
```
We don't have error handling here. If the `getAvailableFetcherOptions` call fails, the dialog is constantly in loading mode.

luca-schlecker replied:
Should display an error now.

domienderle commented on src/lib/components/composites/settings/project-settings/slr/FetcherOptionsDialog.svelte:
```diff
@@ -0,0 +1,156 @@
+    let loading = $state(false);
+
+    const headers: [string, string][] = [
+        ["Overridden", "Indicates wether the fetcher receives this option."],
```
```suggestion
        ["Overridden", "Indicates whether the fetcher receives this option."],
```

domienderle commented on src/lib/components/composites/settings/project-settings/slr/FetcherRemovalDialog.svelte:
```diff
@@ -0,0 +1,59 @@
+        delete updatedFetcherApis[fetcher];
+        await updateFetchers(projectId, updatedFetcherApis, onProjectChanged, (it) => (error = it));
+        loading = false;
+        open = false;
```
Add error handling here. It is an error shown, if the call fails, but then the dialog is directly closed. I would expect, that the dialog stas opened, but with the according error.

luca-schlecker replied:
Should be resolved now

domienderle commented on src/lib/components/composites/settings/project-settings/slr/FetcherSettings.svelte:
```diff
@@ -0,0 +1,143 @@
+        projectSettings = project.settings;
+        availableFetchers = await backendService
+            .getAvailableFetchers({})
+            .then((it) => it.response.fetcherNames);
```
Error handling?

luca-schlecker replied:
Should be in place now.

domienderle commented on src/lib/components/composites/settings/project-settings/slr/FetcherSettings.svelte:
```diff
@@ -0,0 +1,143 @@
+
+    // Fetch project (projectId) from backend and update current state with it
+    async function refetchProject() {
+        loading = true;
```
add `error = undefined;`. Otherwise the old errors are still visible.

domienderle commented on src/lib/components/composites/settings/project-settings/slr/FetcherSettings.svelte:
```diff
@@ -0,0 +1,143 @@
+    let optionDialogOpen = $state(false);
+    let addDialogOpen = $state(false);
+    let removalDialogOpen = $state(false);
+    let selectedFetcherForOptionEditing: string = $state("");
+    let selectedFetcherForRemoval: string = $state("");
```
The variable names are very long. Maybe rename into `fetcherToEdit` and `fetcherToRemove`

luca-schlecker replied:
Great suggestions! Thank you ^^

domienderle commented on src/lib/components/composites/settings/project-settings/slr/FetcherSettings.svelte:
```diff
@@ -0,0 +1,143 @@
+        <Alert details={error.errorDetails} title={error.errorTitle} variant="error" />
+    {/if}
+    {#if usedFetchers}
+        {#if usedFetchers.length == 0}
```
```suggestion
        {#if usedFetchers.length === 0}
```

domienderle reviewed (changes requested):
First of all, thanks for the adjustments. Mostly the feature looks great now. I really like the loading spinner at the section label of the fetchers. Maybe we can discuss about adding such labels in the other settings sections too at some point. Even though I've still discovered a few edge cases (to be honest some of them are really only edge cases, but nevertheless imo they should be covered:

- If I click on the delete button in the delete dialog (loading spinner occurs), then click the cancel button it is possible to click the delete button (normal not in the dialog) again. Maybe, you can disable the cancel button as long as the fetcher gets deleted (this also refers to the cancel button in the add fetcher dialog, because the problem here is of similar nature)
- The save button if you edit the option values, is only a loading button if you open the edit option values. If you click on save doesn't show a loading spinner and is still enabled (disable the cancel button here as well, until the save is done)
- Should it be possible to save empty string as option values? At the moment this is possible and I don't know any useful case for that, right?
- If we deactivate the mock-backend and try to add a fetcher, then no error is shown, but nothing happens, indicating that everything worked, which was not the case. Please correct the error handling here.


Note: I added a bug report, because the open button if you created a new project is not a loading spinner (see #489). You don't have to, but if you find time to also include this in this PR it would be great. It should not be much effort to do this. But it's also fine, if you leave it for later, I just wanted to mention it.

> 1. It is a finesse of how options are handled: The overridden column indicates if the option is actually sent to the fetcher. So even if you insert the displayed default, it is still in fact sent to the fetcher.

All right. This make sense in some way. The question here is, if it is useful, because imo the user doesn't matter if the value is sent to the fetcher or not, right? But we can also discuss this in the channel with the whole team, if you want to.

> 2. The order is dictated by the maps involved in the process. It is possible that you only noticed it that heavily because all fetchers in the mock backend share the same set of options, making a duplicate encounter statistically quite likely. I'd like to leave it like this for now and react in the future if it turns out to be an annoyance.

All right. Thanks for the clarification. Then we keep it as it is for now.

I did not have a look at your tests right now. I will do this, as soon as we are mostly done with the rest. But please add integration tests for your components as well.

luca-schlecker committed:
feat: address pr feedback

domienderle commented on tests/e2e/project/settings/slr/fetcher-options.test.ts:
```diff
@@ -0,0 +1,72 @@
+        await slrPage.insertFetcher("test");
+        await reloadWait(slrPage.page, slrPage.getFetcherTitle("test"));
+        await slrPage.openEditFetcherDialog("test");
+        await expect(slrPage.page.getByRole("checkbox").nth(0)).not.toBeChecked();
```
Use `first()` instead of `nth(0)`. Please correct this every time you use `nth(0)` here and in `project-slr-settings-page-model.ts` as well.
```suggestion
        await slrPage.page.getByRole("textbox").first().fill("foobar");
```

domienderle commented on tests/e2e/project/settings/slr/project-slr-settings-page-model.ts:
```diff
@@ -45,4 +50,51 @@ export class ProjectSLRSettingsPageModel {
         }
     }
+
+    async insertFetcher(fetcherName: string) {
+        await this.apiClient.updateProject({
+            project: Project.create({
+                id: this.projectId,
+                settings: Project_Settings.create({
+                    fetchers: Object.fromEntries([[fetcherName, {}]]),
+                }),
+            }),
+            mask: {
+                paths: ["settings.fetchers.test"],
+            },
+        });
+    }
```
Since we want to test the end-to-end behavior of the website don't call the backend methods yourself. The tests really should simulate the user behavior. Please remove this and use `addFetcher` instead.

domienderle commented on tests/e2e/project/settings/slr/fetcher-options.test.ts:
When you have removed the `insertFetcher` and replaced it with add fetcher, you can also think of moving the `addFetcher` into the fixture, since you use it in every test.

luca-schlecker replied:
It is already in the page-model.

domienderle replied:
Yes it also good there. Fine for me.

domienderle reviewed (commented):
Great work! Now there are only a couple of things left. Please answer every comment and every question. Also please fix the e2e test, so they work and as I mentioned in the last review as well, please add integration tests for your components too:

> If I click on the delete button in the delete dialog (loading spinner occurs), then click the cancel button it is possible to click the delete button (normal not in the dialog) again. Maybe, you can disable the cancel button as long as the fetcher gets deleted (this also refers to the cancel button in the add fetcher dialog, because the problem here is of similar nature)

This is still possible. I can still click the cancel button and then the delete button as long as the fetcher gets deleted. Have a look on the other buttons as well, because the behavior here is the same.

> Should it be possible to save empty string as option values? At the moment this is possible and I don't know any useful case for that, right?

Please answer on this question.

luca-schlecker commented:
> Great work! Now there are only a couple of things left. Please answer every comment and every question. Also please fix the e2e test, so they work and as I mentioned in the last review as well, please add integration tests for your components too:
> 
> > If I click on the delete button in the delete dialog (loading spinner occurs), then click the cancel button it is possible to click the delete button (normal not in the dialog) again. Maybe, you can disable the cancel button as long as the fetcher gets deleted (this also refers to the cancel button in the add fetcher dialog, because the problem here is of similar nature)
> 
> This is still possible. I can still click the cancel button and then the delete button as long as the fetcher gets deleted. Have a look on the other buttons as well, because the behavior here is the same.
> 
> > Should it be possible to save empty string as option values? At the moment this is possible and I don't know any useful case for that, right?
> 
> Please answer on this question.

I didn't mean to already request the review, hence why I instantly unrequested it. Sorry for the inconvenience.

luca-schlecker committed:
test: use more widely available docker syntax for ephemeral ports

luca-schlecker committed:
feat(FetcherOptions): always disable cancel button when loading

luca-schlecker committed:
test(FetcherAddDialog): add integration test

luca-schlecker committed:
test: use mock backend version v7

luca-schlecker committed:
feat: miscellaneous fixes

domienderle commented on tests/integration/settings/project-settings/slr/fetcher-add-dialog.test.ts:
Please split this in the following tests:
- The component with fetchers is rendered correctly
- The fetcher adding works correctly
This shoould mostly be a refactoring, but it makes the tests more readable.

domienderle commented on tests/integration/settings/project-settings/slr/fetcher-option-row.test.ts:
```diff
@@ -0,0 +1,56 @@
+        expect(checkBox).not.toBeChecked();
+    });
+
+    test("When you enter something, then the value is not overridden", async () => {
```
```suggestion
    test("When you enter something, then the value is overridden", async () => {
```

domienderle commented on tests/integration/settings/project-settings/slr/fetcher-option-row.test.ts:
Please add a test if the component is rendered correctly.

domienderle commented on tests/integration/settings/project-settings/slr/fetcher-options-dialog.test.ts:
Same here as for the `FetcherAddDialog`. Please split the test into the following:
- Component renders correctly
- Functionality works correct

domienderle commented on tests/integration/settings/project-settings/slr/fetcher-removal-dialog.test.ts:
```diff
@@ -0,0 +1,60 @@
+    beforeEach(() => vi.clearAllMocks());
+    afterAll(() => vi.restoreAllMocks());
+
+    test("When a fetcher is deleted, then it is deleted", async () => {
```
```suggestion
    test("When a fetcher is deleted, then it is deleted correctly", async () => {
```

domienderle commented on tests/setupTest.ts:
Same here. Please have two tests:
- Component rendering works
- Functionality works

domienderle reviewed (changes requested):
Good work. Only a few things left now:

- If you add a fetcher and cancel, then it waits until all is loaded correctly whereas it is still possible to click the cancel button and afterwards the delete oder edit button again (even already takes longer than before - good work - but anyway it is possible which should not be the case. 
-    > Should it be possible to save empty string as option values? At the moment this is possible and I don't know any useful case for that, right?

     Please answer this question

-  I don't know how the currents status about the API versions is. Why do you update the API version instantly to v7.0.0 here, instead of v6.0.0 as it was before in this PR?

domienderle commented:
Additionally, please have a look on the `Teamscale` findings.

luca-schlecker committed:
feat: address pr feedback

luca-schlecker commented:
> Good work. Only a few things left now:
> 
> * If you add a fetcher and cancel, then it waits until all is loaded correctly whereas it is still possible to click the cancel button and afterwards the delete oder edit button again (even already takes longer than before - good work - but anyway it is possible which should not be the case.
> * > Should it be possible to save empty string as option values? At the moment this is possible and I don't know any useful case for that, right?
>   
>   
>   Please answer this question
> * I don't know how the currents status about the API versions is. Why do you update the API version instantly to v7.0.0 here, instead of v6.0.0 as it was before in this PR?

1. I resolved it by showing skeletons instead of the old fetchers when updating.
2. I think allowing empty boxes and spaces is fine. This should just forward the data and the plugin then decides what should happen (trimming and co.).
3. I have not updated the version to v0.7.0. I've double-checked and also reinitialized the repository submodules, it should be at v0.6.0.

luca-schlecker commented:
> Additionally, please have a look on the `Teamscale` findings.

I had a look but don't think any of them are of relevance. Test methods can be long, and callback types should have a name (which is seen as being unused).

domienderle reviewed (dismissed):
All right, we are good to go now. Great work. About the `Teamscale` findings, I didn't mean to obligatorily change your code, but tolerate them, if you decide that it is ok to have them, so we are able to merge.

> I resolved it by showing skeletons instead of the old fetchers when updating.

Good works. Thanks.

> I think allowing empty boxes and spaces is fine. This should just forward the data and the plugin then decides what should happen (trimming and co.).

All right. That's fine for me then.

> I have not updated the version to v0.7.0. I've double-checked and also reinitialized the repository submodules, it should be at v0.6.0.

Sorry, you are right. I got something wrong.

luca-schlecker commented:
Rebased this :+1:

domienderle reviewed (changes requested):
Sorry, that this PR takes so long. I have two more additions, but then we are done:

- I recognized, that if the SLR settings are locked (see the `MaybeAsDecisionButton`) all SLR settings should be locked. Please add this functionality, that for example if at least one review is already submitted the fetcher options are disabled.

- Can you increase the gap between the fetcher options and the section above a little bit, so that the sections are clearly separated.

Additionally, tolerate the `Teamscale` findings, before we can merge this.

luca-schlecker committed:
feat: increase gap in project settings layout

luca-schlecker commented:
> Sorry, that this PR takes so long. I have two more additions, but then we are done:
> 
> * I recognized, that if the SLR settings are locked (see the `MaybeAsDecisionButton`) all SLR settings should be locked. Please add this functionality, that for example if at least one review is already submitted the fetcher options are disabled.
> * Can you increase the gap between the fetcher options and the section above a little bit, so that the sections are clearly separated.
> 
> Additionally, tolerate the `Teamscale` findings, before we can merge this.

I just tested it manually, and I am unsure if locking these controls is even necessary. I was not able to navigate to the SLR settings of an already locked project, neither through the button nor through manually typing the URL. Could you please verify that on your side too? If this is the case, I'd prefer to leave it out, as the case is already handled.

Furthermore, I've increased the gap and fixed overflow scrolling.

luca-schlecker committed:
feat: allow "disabled" in and add action button icon to AlertDialog

luca-schlecker committed:
fix: slrSettingsLocked now locks fetcher options too

domienderle reviewed (approved):
All right. Now we are ready to merge. Good work.
